toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -89,41 +94,196 @@ auto GenerateRandomIdentifier(absl::BitGen& gen, int min_length = 1,
   table.push_back(TokenKind::TokenName);
 #define CARBON_OPENING_GROUP_SYMBOL_TOKEN(TokenName, Spelling, ClosingName)
 #define CARBON_CLOSING_GROUP_SYMBOL_TOKEN(TokenName, Spelling, OpeningName)
 #include "toolchain/lexer/token_kind.def"
-  table.insert(table.end(), 32, TokenKind::Semi);
-  table.insert(table.end(), 16, TokenKind::Comma);
-  table.insert(table.end(), 12, TokenKind::Period);
-  table.insert(table.end(), 8, TokenKind::Colon);
-  table.insert(table.end(), 8, TokenKind::Equal);
-  table.insert(table.end(), 4, TokenKind::Amp);
-  table.insert(table.end(), 4, TokenKind::ColonExclaim);
-  table.insert(table.end(), 4, TokenKind::EqualEqual);
-  table.insert(table.end(), 4, TokenKind::ExclaimEqual);
-  table.insert(table.end(), 4, TokenKind::MinusGreater);
-  table.insert(table.end(), 4, TokenKind::Star);
-  return table;
-}
-
-auto GetSymbolTokenTable() -> llvm::ArrayRef<TokenKind> {
-  static auto symbol_token_table_storage = GetSymbolTokenTableImpl();
+    table.insert(table.end(), 32, TokenKind::Semi);
+    table.insert(table.end(), 16, TokenKind::Comma);
+    table.insert(table.end(), 12, TokenKind::Period);
+    table.insert(table.end(), 8, TokenKind::Colon);
+    table.insert(table.end(), 8, TokenKind::Equal);
+    table.insert(table.end(), 4, TokenKind::Amp);
+    table.insert(table.end(), 4, TokenKind::ColonExclaim);
+    table.insert(table.end(), 4, TokenKind::EqualEqual);
+    table.insert(table.end(), 4, TokenKind::ExclaimEqual);
+    table.insert(table.end(), 4, TokenKind::MinusGreater);
+    table.insert(table.end(), 4, TokenKind::Star);
+    return table;
+  }();
   return symbol_token_table_storage;
 }
 
-// Generate random symbols. This skews the distribution as best it can towards
-// what we expect in real world source code, but doesn't include grouping
-// symbols for simplicity.
-auto GenerateRandomSymbol(absl::BitGen& gen) -> llvm::StringRef {
-  llvm::ArrayRef<TokenKind> table = GetSymbolTokenTable();
-  auto index = absl::Uniform<int>(gen, 0, table.size());
-  return table[index].fixed_spelling();
+// Compute a random sequence of mixed symbols, keywords, and identifiers, with
+// percentages of each according to the parameters.
+auto RandomMixedSeq(int symbol_percent, int keyword_percent) -> std::string {
+  std::string result;
+  absl::BitGen gen;
+
+  // Get static pools of symbols, keywords, and identifiers.
+  llvm::ArrayRef<TokenKind> symbols = GetSymbolTokenTable();
+  llvm::ArrayRef<TokenKind> keywords = TokenKind::KeywordTokens;
+  llvm::ArrayRef<std::string> ids = GetRandomIdentifiers();
+
+  // Build a list of kind keys and indices into the relevant tables that have
+  // the desired distribution, then shuffle that list.
+  enum ElementKind {
+    Symbol,
+    Keyword,
+    Identifier,
+  };
+  std::array<std::pair<ElementKind, int>, NumTokens> indices;
+  for (int i : llvm::seq(0, NumTokens)) {
+    int i_percent = i % 100;
+    if (i_percent < symbol_percent) {
```

Comment:
1. I think you might get an odd distribution here because of how you split by i_percent, then do the i %. You get non-sequential `i` values that you then mod.

2. Also, you do `i_percent < symbol_percent` and `i_percent < keyword_percent`. With symbol_percent = 50 and keyword_percent = 20, doesn't that result in no keywords?

Had you considered something like loops for each:

```
CARBON_CHECK(symbol_percent + keyword_percent < 100) << "These are percents, and to avoid handling rounding errors we expect at least some identifiers.";
int symbol_count = NumTokens * 100 / symbol_percent;
for (int i : llvm::seq(0, symbol_count)) {
  indices[i] = {Symbol, i % symbols.size()};
}
int keyword_count = NumTokens * 100 / keyword_percent;
for (int i : llvm::seq(symbol_count, symbol_count + keyword_count)) {
  indices[i] = {Keyword, i % keywords.size()};
}
for (int i : llvm::seq(symbol_count + keyword_count, NumTokens)) {
  // Note as currently implemented this doesn't actually need the mod, since ids.size() == NumTokens.
  // But I'm wondering if that should be fixed.
  indices[i] = {Identifier, i % ids.size()};
}
```

Reply:
Oof, yeah, all of this logic was just busted. Reworking it along the lines of your suggestion.

That said, I think its fine to support 0% identifiers here, so I'll let these sum right up to 100. That seem OK?